### ISSUE TITLE
Feature Request: Per-directory configuration

### ISSUE THREAD

**Issue by anomiex** [PHPCSStandards/PHP_CodeSniffer#126]
## Is your feature request related to a problem?

It's difficult to write and maintain XML configuration rules to apply different subdirectories, as the existing configuration file syntax is oriented towards specifying paths that apply to rules rather than rules that apply to paths.

See also https://github.com/squizlabs/PHP_CodeSniffer/issues/2551

## Describe the solution you'd like

Per-directory configuration files would be nice. Conceptually there are at least two ways this might work: for example, if you have a file at `subdir/.phpcs.dir.xml` then for files in that directory phpcs would

1. act as if it had been run as `phpcs --standard=.phpcs.xml.dist,subdir/.phpcs.dir.xml`, loading the subdirectory ruleset as a peer of the parent's, or
2. use `subdir/.phpcs.dir.xml` as the ruleset, but treat that file as if it had `<rule ref="../.phpcs.xml.dist" />` at the start.

## Additional context (optional)

We've been using [a fork of phpcs](https://github.com/Automattic/PHP_CodeSniffer/) that enables this in our monorepo at https://github.com/Automattic/jetpack since December 2021 and it has worked out very well for us despite some caveats.

The important caveats with the version in the fork are:
* The fork itself is pretty minimal, as in the old repo it seemed like 4.0 was never going to happen. Basically the fork just enables filters to return a `PHP_CodeSniffer\File` instance instead of a filename. Then a custom filter returns File instances that have different values for `->ruleset` and `->config` as necessary to reflect the per-dir config files.
* Since the `PHP_CodeSniffer\Config` class is half-static, `<config>` doesn't work well at a per-directory level.
  * Various other directives like `<ini>` and `<arg>` and `<autoload>` also don't work at a per-directory level, but those are much easier to classify as "global" and therefore say they shouldn't work per-directory.
* Since our fork takes the `--standard=.phpcs.xml.dist,subdir/.phpcs.dir.xml` approach rather than the implicit-`<rule ref="../.phpcs.xml.dist" />` approach (because the latter is a lot harder to do from a filter), `<exclude name="Standard.Category.Rule" />` doesn't work to exclude a rule that had been enabled in a parent directory. Since `<exclude name="Standard.Category.Rule"><severity>0</severity></exclude>` does still work, that hasn't been that much of a problem.

Since a 4.0 release seems imminent in the new repo, I wonder whether I might do better to do a more invasive PR than we had done in https://github.com/squizlabs/PHP_CodeSniffer/pull/3378. The main breaking change a more invasive PR would do would be to change methods like `Config::getConfigData()` to be non-static, which would affect sniffs that use Config. Probably they'd have to access the instance via `$phpcsFile->config` instead.

- [x] I intend to create a pull request to implement this feature.

**Comment by jrfnl** [comment]
I'm not sure I fully understand your problem.

You can already use option 2 by having an explicit `<rule ref="../.phpcs.xml.dist" />` at the start of the subdirectory ruleset.
Making that implied will be a huge breaking change which I'm not willing to make.

Having it explicit is a one-liner, so why is anything else needed ? (aside from the overruling issue, but that would be solved once 4.0 is released).


> Since a 4.0 release seems imminent in the new repo, I wonder whether I might do better to do a more invasive PR than we had done in https://github.com/squizlabs/PHP_CodeSniffer/pull/3378. The main breaking change a more invasive PR would do would be to change methods like Config::getConfigData() to be non-static, which would affect sniffs that use Config. Probably they'd have to access the instance via $phpcsFile->config instead.

That is not a change I would accept for the 4.0 release. To be able to get to 4.0 sooner rather than later, I want to _stabilize_ things, not de-stabilize by making larger breaking changes for which I cannot oversee the full impact of at this time.

**Comment by anomiex** [comment]
I think the part you're missing is that I want to run `phpcs` just once at the root of the project, and then for `some/path/` that has `some/path/.phpcs.dir.xml` it would transparently use that standard for files under that path, and for `another/path/` with a different `another/path/.phpcs.dir.xml` it would use that standard for files in there, and for `another/path/and/deeper/` with `another/path/and/deeper/.phpcs.dir.xml` it would use that standard (which itself "extends" `another/path/.phpcs.dir.xml`) for files there.

Yes, you can get something close to this with configuration like you suggest _and_ multiple runs of phpcs _and_ careful exclusion of the files in the higher-level runs that will be covered in the subdirectory runs instead. And then possibly having to merge together the results from the multiple runs for CI tooling to make use of. But all that is a pain and hard to scale.

Note the implied `<rule ref="../.phpcs.xml.dist" />` would only apply to these per-directory `.phpcs.dir.xml` config files. If someone is already using the multiple-runs method instead, those wouldn't be per-directory config files and so wouldn't have the implied `<rule>`. And I'm still leaning towards there being no default name for the per-directory files, so anyone wanting to use it would have to opt-in by setting a name for them in their configuration.

I do like the implied `<rule>` for the per-dir files, because then you don't have to worry keeping them in sync when changing things around. Say you were to decide to add `another/.phpcs.dir.xml`, for example: with an explicit `<rule>` you'd have remember to update `another/path/.phpcs.dir.xml` (and any others) to point to that now instead of the `.phpcs.xml.dist` in the root.

> That is not a change I would accept for the 4.0 release.

Thanks for letting me know. Would 5.0 be expected to happen reasonably soon after 4.0 (like months to a year rather than multiple years) and a breaking PR would be accepted for that? Or would you rather have a rebase of https://github.com/squizlabs/PHP_CodeSniffer/pull/3378, since that was carefully designed to not break anything?

**Comment by jrfnl** [comment]
> I think the part you're missing is that I want to run `phpcs` just once at the root of the project, and then for `some/path/` that has `some/path/.phpcs.dir.xml` it would transparently use that standard for files under that path, and for `another/path/` with a different `another/path/.phpcs.dir.xml` it would use that standard for files in there, and for `another/path/and/deeper/` with `another/path/and/deeper/.phpcs.dir.xml` it would use that standard (which itself "extends" `another/path/.phpcs.dir.xml`) for files there.

Indeed the bit I was missing. Will need to have a think about this, but this is not something which is on my priority list at this time.

> > That is not a change I would accept for the 4.0 release.
> 
> Thanks for letting me know. Would 5.0 be expected to happen reasonably soon after 4.0 (like months to a year rather than multiple years) and a breaking PR would be accepted for that?

I do expect to have more regular majors, once a year/once every two years.

However, my focus at this time is on stabilizing and on getting to 4.0 and then on some structural changes I already have planned for 5.0 myself. I currently don't have the bandwidth to make a judgement call on this ticket for 5.0. I will need to circle back to this at a later time.

**Comment by paulshryock** [comment]
> It's difficult to write and maintain XML configuration rules to apply different subdirectories, as the existing configuration file syntax is oriented towards specifying paths that apply to rules rather than rules that apply to paths.

Is this not a problem that could be solved in the main configuration file without needing to have additional configuration files?

I feel like a single configuration file is ideal.

If the current config file format doesn't provide enough flexibility for customization in the way that's desired, maybe that's something that could be improved?

The other proposed solutions (per-directory configuration files) feel like a lot of extra unnecessary complexity.

**Comment by jrfnl** [comment]
> Is this not a problem that could be solved in the main configuration file without needing to have additional configuration files?

For the most part, but not if a projects wants to use different value for certain sniff properties for specific subdirectories.
Sniff properties can only have one value for a run.

**Comment by anomiex** [comment]
> I feel like a single configuration file is ideal.

The eslint developers thought the same thing when they first added their "flat" configuration in v9. Then they [brought back having configuration in subdirectories](https://eslint.org/docs/latest/use/migrate-to-10.0.0#-new-configuration-file-lookup-algorithm) because it turns out that managing all your various subproject configurations for a monorepo in one file can get confusing and hard to manage.

And PHPCS's format would be even harder to manage, since you'd have to repeat the `<exclude-pattern>` and `<include-pattern>` subelements in all the relevant `<rule>` elements instead of wrapping different blocks of `<rule>`s.

In our monorepo, where we've been using a forked version of phpcs with this feature for a long time, we've found it very convenient to be able to do things like drop a small file enabling a testing-file-specific ruleset in `tests/` dirs instead of having to enumerate all the `tests/` dirs in one global config file.

> Sniff properties can only have one value for a run.

It _is_ possible (at least in v3) to have sniff properties vary for different files. The sniff properties are non-static fields on the `Sniff` instances, `Sniff`s are accessed via the `File` instance's `Ruleset` instance, and it works fine for different `File` instances to point to different `Ruleset` instances.

OTOH, `<config>` settings do need to have one value for the whole run: even though `LocalFile` also points to a `Config` instance, `Config::getConfigData()` and the like are static, so the different instances all access the same configuration data. Which is unfortunate for us, as we have some dirs that need `<config name="testVersion" value="7.2-"/>` while others could do better to specify later versions.

**Comment by renatonascalves** [comment]
In a recent PHP conference in Brazil, I was asking @rodrigoprimo about a lack of “monorepo support” for this project. And he brought this issue to my attention.

I’d like to share _briefly_ how we are using PHP_CodeSniffer in a big monorepo. Interestingly, the things I wished existed is similar to the @anomiex wishlist, but we are approaching them differently.

----

TL;DR: we have a monorepo that’s used by multiple brands. So each brand has a child theme that extend a parent theme. This parent theme is considered “global” code (shared by all brands). Each brand has their own teams. There are also “global”/shared plugins, and per brand plugins.

The way we run `phpcs` is from the root of the monorepo. Both locally and CI (including via pre-commit checks only for modified files but using the root config). We don’t have custom configs per dir. We have a single root config, and from this config we say which rules dirs should be excluded globally. Which rules are disabled per dirs from the root, etc.

Having used the one config per-dir approach before (even before migrating to the current config organization), I’ve found the approach of having a root config to be way simpler, faster, and maintenance friendly. It’s just easier to everybody to reason about (I’ve found at least), particularly with many different teams working on a big codebase.

Could we have per dir configs? Yes. But I feel there is something cluncky with this way of running it. It can take longer for CI to run (even though I’d agree the performance impact is highly debatable). It is harder view errors (since they are presented per dir). And it creates a need to “copy duplicate rules” all of the place. Where most of them are just “importing” the main config from the parent. One starts to question the usefulness of per dir config when you are not actually creating per-dir custom rules.

Two recent scenarios came to mind for my comment about a lack of “monorepo support”. The first scenario I run into is that you can’t set different text-domains from a global config to specific paths (sub-directories). Someone already created an issue [here](https://github.com/WordPress/WordPress-Coding-Standards/issues/1650), but as far as I understand, it’d need changes in this repo to make it happen.

### Different text-domains per subdirectory

E.g: in my monorepo, we have a global text-domain, and a text-domain per brand. Not per theme or plugin. Each brand have their own text-domain. So, one thing `phpcs` can’t do today is: a developer working on a plugin from brand A, and by mistake, adds text-domain from brand B, `phpcs` can’t report that mistake.

PHPCS only knows that text-domains from brand A and B exist, and that they can be used in the whole monorepo. Not that one should not use one in specific dirs (brand-specific themes and plugins). Only a per-dir config could do this today, which we don’t use.

So this feature today is lacking in our monorepo. I’ve yet to carve out time to try a custom sniff to try to accomplish this.

### DiscouragedFunctionsSniff per subdirectory

Another scenario that’s frequent is about using `DiscouragedFunctionsSniff`. We can use it to discourage usage of a specific function and suggest another (for any reason). But imagine you are improving your codebase with a new feature (a new global function) that all brands should start using. And you can’t simply migrate everything at once to this new helper function.

`DiscouragedFunctionsSniff` is a perfect tool for this. Because one could simply block a legacy function from being used per dir (in brand-specific themes first, then plugins, etc). Today, one can’t do that without creating a custom sniff (which is what we did but is not perfect).

There are other similar examples that I don’t remember now, but I can think deeper if needed. But those two examples is enough to share my point.

### Some final thoughts

Having said all that, and while looking at some other issues and prs here, e.g: [1](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/126), [2](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/380), [3](https://github.com/squizlabs/PHP_CodeSniffer/pull/3378), [4](https://github.com/squizlabs/PHP_CodeSniffer/issues/2551). I do think “monorepo support” is tricky because apparently some people want to manage per dir configs and override those rules there “automatically”. And others, like me, wants root configs but with more per-dir rules flexibility.

Obviously, I don’t know the project technically to say if both can be satisfied. But I’m leaving this feedback here, just to highlight that whatever route the project takes, having monorepo support is very useful. It’s possible it might be niche-based too. 

:)

**Comment by kasparsd** [comment]
I got [nerd-snipped](https://xkcd.com/356/) by the blog post from @renatonascalves and created this phpcs wrapper package https://github.com/wpelevator/phpcs-parallel which runs `phpcs` and `phpcbf` across any number of sub-projects, for example:

```bash
phpcs-parallel --config-pattern='packages/*' --processes=4
```

will run `phpcs --standard=packages/*/phpcs.xml.dist packages/*` for each directory with `phpcs.*` config under the `packages` directory. The nice side effect of this is that this allows parallel runs which improves the performance significantly in larger codebases.

Maybe that's a useful way to test the idea and polish the DX before proposing a more generic solution for the upstream project.
